README.md
# RGB REST JavaScript Client

A JavaScript library used to consume the RGB REST API service provided by Blockchain of Things.

## Installation

On Node.js:

```shell
npm install rgb-rest-client
```

On the browser:

```html
<script src="https://unpkg.com/rgb-rest-client"></script>
```

### Browser compatibility

The RGB REST JavaScript Client library is compatible with modern web browsers.

It has been tested on the following web browsers:

- Safari ver. 14.0 (on macOS Big Sur 11.2)
- Google Chrome ver. 89.0 64 bits (on macOS Big Sur 11.2)
- Firefox ver. 86.0 64-bits (on macOS Big Sur 11.2)
- Microsoft Edge ver. 89.0 64 bits (on macOS Big Sur 11.2)

## Usage

### Instantiate the RGB REST client object.

On Node.js:

```javascript
const rgbRestClient = require('rgb-rest-client');

const rgb = new rgbRestClient.RgbRestClient();
```

On the browser:

```html
<script>
const rgb = new rgbRestClient.RgbRestClient();
</script>
```

#### Constructor options

The following options can be used when instantiating the RGB REST client object:

- **network** \[String\] - (optional, default: <b>*'testnet'*</b>) Target Bitcoin blockchain network. Valid values:
    - *'testnet'*
    - *'signet'*
- **host** \[String\] - (optional, default: <b>*'rgb.blockchainofthings.com'*</b>) Host name (with optional port) of target RGB REST API server.
- **secure** \[String\] - (optional, default: <b>*true*</b>) Indicates whether a secure connection (HTTPS) should be used.
- **version** \[String\] - (optional, default: <b>*'1.1'*</b>) Version of RGB REST API to target.

### Calling API methods

#### Issue fungible asset

Using a callback:

```javascript
rgb.issueFungible(
    'TST1',           // Ticker
    'Test asset #1',  // Name
    0,                // Precision
    {                 // Allocations
        coins:123,
        outpoint:'5c1cb187a0d80955f87d65e3b798b5975362aa5e6f48d99a293bf162606fbdaa:4'
    },
    {                 // Options
        description:'RGB asset used for testing RGB REST client'    // Description
    },
    (err, data) => {
        if (err) {
            console.error('Error issuing fungible RGB asset:', err);
        }
        else {
            console.log('Newly issued asset info:', data.assetInfo);
        }
    }
);
```

Using promise:

```javascript
rgb.issueFungible(
    'TST1',           // Ticker
    'Test asset #1',  // Name
    0,                // Precision
    {                 // Allocations
        coins:123,
        outpoint:'5c1cb187a0d80955f87d65e3b798b5975362aa5e6f48d99a293bf162606fbdaa:4'
    },
    {                 // Options
        description:'RGB asset used for testing RGB REST client'    // Description
    }
)
.then(data => {
    console.log('Newly issued asset info:', data.assetInfo);
})
.catch(err => {
    console.error('Error issuing fungible RGB asset:', err);
});
```

#### Retrieve unspent Bitcoin transaction outputs managed by a wallet

Using a callback:

```javascript
rgb.walletUtxos(
    'wpkh([bbdc7dee/0\'/0\']tpubDAi9mxwvCucazfwjsVTe7BQ9imvKsLA2ZifaHpETwAEtMNroXkWeVAQGhq6EyhD7i8fBCE2mVaeLL638g8zD1faLXtcFHsvAdxuD1ffNgWs/*\')#nx3xqje2', // Wallet descriptor
    {       // Options
        keyRangeStartIdx: 0,
        keyRangeCount: 100
    },
    (err, data) => {
        if (err) {
            console.error('Error retrieving wallet UTXOs:', err);
        }
        else {
            console.log('Wallet UTXOs:', data.utxos);
        }
    }
);
```

Using promise:

```javascript
rgb.walletUtxos(
    'wpkh([bbdc7dee/0\'/0\']tpubDAi9mxwvCucazfwjsVTe7BQ9imvKsLA2ZifaHpETwAEtMNroXkWeVAQGhq6EyhD7i8fBCE2mVaeLL638g8zD1faLXtcFHsvAdxuD1ffNgWs/*\')#nx3xqje2', // Wallet descriptor
    {       // Options
        keyRangeStartIdx: 0,
        keyRangeCount: 100
    }
)
.then(data => {
    console.log('Wallet UTXOs:', data.utxos);
})
.catch(err => {
    console.error('Error retrieving wallet UTXOs:', err);
});
```

### Error handling

Two types of error can take place when calling API methods: client or API error.

Client errors return generic error objects.

API errors, on the other hand, return a custom **RgbRestError** object.

**RgbRestError** objects are extended from Javascript's standard *Error* object, so they share the same
characteristics with the following exceptions:

- The value of the *name* field is set to `RgbRestError`
- It has the following additional fields: *httpStatusMessage*, *httpStatusCode*, and *apiErrorMessage*

> **Note**: the *apiErrorMessage* field of the RgbRestError object contains the error message returned by the
RGB REST API service. However, there might be cases where that field is **undefined**.

Usage example:

```JavaScript
const rgbRestClient = require('rgb-rest-client');

const rgb = new rgbRestClient.RgbRestClient();

rgb.issueFungible(
    '',     // Ticker
    '',     // Name
    0,      // Precision
    [],
    (err, data) => {
        if (err) {
            if (err instanceof rgbRestClient.RgbRestError) {
                // RGB REST API error
                console.log('HTTP status code:', err.httpStatusCode);
                console.log('HTTP status message:', err.httpStatusMessage);
                console.log('API error message:', err.apiErrorMessage);
                console.log('Compiled error message:', err.message);
            }
            else {
                // Client error
                console.log(err);
            }
        }
        else {
            // Process returned data
        }
    }
);
```

Expected result:

```
HTTP status code: 400
HTTP status message: Bad Request
Catenis error message: Invalid parameters
Compiled error message: Error returned from RGB REST API endpoint: [400] Invalid parameters
```

## License

This JavaScript library is released under the [MIT License](LICENSE). Feel free to fork, and modify!

Copyright © 2021, Blockchain of Things Inc.


CHANGELOG.md
# Changelog

## [1.1.3] - 2021-04-05

### Changes
- Corrected comments that document the 'issueFungible' method.

## [1.1.2] - 2021-04-03

### Changes
- Updated Readme file.
- Added Changelog file.

## [1.1.1] - 2021-04-03

### Fixes
- Corrected the version of the RGB REST API that is targeted by this library.

## [1.1.0] - 2021-04-03

### New features
- Added support for new RGB REST API method "Wallet unspent transaction outputs".

## [1.0.0] - 2021-03-16

### New features
- Initial version targeting version 1.0 of the RGB REST API.

test/testRgbRestClient.js
/**
 * Created by claudio on 2021-03-09
 */

const rgbRestClient = require('../src/index');
const expect = require('chai').expect;

const testSuite = require('./suite/RgbRestClientSuite');

testSuite(rgbRestClient, expect);


test/browser/runTestSuite.js
/**
 * Created by claudio on 2021-03-09
 */

this.testSuite(this.rgbRestClient, this.chai.expect);


test/umd/testRgbRestClient.js
/**
 * Created by claudio on 2021-03-09
 */

const rgbRestClient = require('../../umd/rgb-rest-client.js');
const expect = require('chai').expect;

const testSuite = require('../suite/RgbRestClientSuite');

testSuite(rgbRestClient, expect);


src/index.js
/**
 * Created by claudio on 2021-03-08
 */

const fetch = require('cross-fetch');
const RgbRestError = require('./RgbRestError');

const apiPath = '/api/';

/**
 * RGB REST client class.
 */
class RgbRestClient {
    /**
     * RgbRestClient class constructor.
     * @param {object} [options]
     * @param {string} [options.network=testnet] Target Bitcoin blockchain network.
     * @param {string} [options.host=rgb.blockchainofthings.com] Host name (with optional port) of target RGB REST API server.
     * @param {boolean} [options.secure=true] Indicates whether a secure connection (HTTPS) should be used.
     * @param {string} [options.version=1.1] Version of RGB REST API to target.
     */
    constructor(options) {
        options = options || {};

        this.network = typeof options.network === 'string' && options.network.length > 0 ? options.network : 'testnet';

        let host = typeof options.host === 'string' && options.host.length > 0 ? options.host : 'rgb.blockchainofthings.com';
        let secure = typeof options.secure === 'boolean' ? options.secure : true;
        let version = typeof options.version === 'string' && options.version.length > 0 ? options.version : '1.1';
        let uriPrefix = (secure ? 'https://' : 'http://') + host;
        let apiBaseUriPath = apiPath + version + '/';

        this.rootApiEndPoint = uriPrefix + apiBaseUriPath;
    }

    /**
     * Issue new fungible RGB asset.
     * @param {string} ticker
     * @param {string} name
     * @param {string} precision
     * @param {object|object[]} allocations
     * @param {object} [options]
     * @param {string} [options.description]
     * @param {object|object[]} [options.inflation]
     * @param {string} [options.renomination]
     * @param {string} [options.epoch]
     * @param {function} [callback]
     * @returns {(undefined|Promise)} Undefined if a callback function is passed. Otherwise, a promise that resolves
     *  to an asset info object.
     */
    issueFungible(ticker, name, precision, allocations, options, callback) {
        if (typeof options === 'function') {
            callback = options;
            options = undefined;
        }

        options = options || {};

        let data = {
            ticker,
            name,
            precision,
            allocations
        };

        if (options.description) {
            data.description = options.description;
        }

        if (options.inflation) {
            data.inflation = options.inflation;
        }

        if (options.renomination) {
            data.renomination = options.renomination;
        }

        if (options.epoch) {
            data.epoch = options.epoch;
        }

        return this._request(
            'POST',
            ':network/fungible/issue',
            {
                url: {
                    network: this.network
                }
            },
            data,
            callback
        );
    }

    /**
     * Retrieve unspent Bitcoin transaction outputs (UTXOs) managed by a wallet.
     * @param {string} walletDescriptor
     * @param {object} [options]
     * @param {number} [options.keyRangeStartIdx]
     * @param {number} [options.keyRangeCount]
     * @param {function} [callback]
     * @returns {(undefined|Promise)} Undefined if a callback function is passed. Otherwise, a promise that resolves
     *  to an array of UTXOs.
     */
    walletUtxos(walletDescriptor, options, callback) {
        if (typeof options === 'function') {
            callback = options;
            options = undefined;
        }

        options = options || {};

        const queryParams = {
            walletDescriptor
        };

        if (options.keyRangeStartIdx !== undefined) {
            queryParams.keyRangeStartIdx = options.keyRangeStartIdx;
        }

        if (options.keyRangeCount !== undefined) {
            queryParams.keyRangeCount = options.keyRangeCount;
        }

        return this._request(
            'GET',
            ':network/wallet/utxos',
            {
                url: {
                    network: this.network
                },
                query: queryParams
            },
            callback
        );
    }

    _request(method, methodPath, params, data, callback) {
        if (typeof params === 'function') {
            callback = params;
            params = data = undefined;
        }
        else if (typeof data === 'function') {
            callback = data;
            data = undefined;
        }

        let result;

        // Prepare promise to be returned if no callback passed
        if (typeof callback !== 'function') {
            result = new Promise((resolve, reject) => {
                callback = (err, res) => {
                    if (err) {
                        reject(err);
                    }
                    else {
                        resolve(res);
                    }
                }
            });
        }

        if (method === 'POST') {
            this._postRequest(methodPath, params, data)
            .then(body => callback(null, parseSuccessResponse(body)))
            .catch(err => callback(err));
        }
        else if (method === 'GET') {
            this._getRequest(methodPath, params)
            .then(body => callback(null, parseSuccessResponse(body)))
            .catch(err => callback(err));
        }

        return result;
    }

    async _getRequest(methodPath, params) {
        const url = this._assembleMethodEndPointUrl(methodPath, params);
        const headers = new fetch.Headers();
        headers.append('Content-Type','application/json; charset=utf-8');

        let init = {
            method: 'GET',
            headers
        };

        let res;

        try {
            res = await fetch(url, init);
        }
        catch (err) {
            throw new Error('Error calling GET request: ' + err.toString());
        }

        let body;

        try {
            body = await res.json();
        }
        catch (err) {
            throw new Error('Error reading GET request response: ' + err.toString());
        }

        if (!res.ok) {
            throw new RgbRestError(res.statusText, res.status, typeof body === 'object' && body.message ? body.message : undefined);
        }

        return body;
    }

    async _postRequest(methodPath, params, data) {
        const url = this._assembleMethodEndPointUrl(methodPath, params);
        const headers = new fetch.Headers();
        headers.append('Content-Type','application/json; charset=utf-8');

        let init = {
            method: 'POST',
            headers,
            body: JSON.stringify(data),
        };

        let res;

        try {
            res = await fetch(url, init);
        }
        catch (err) {
            throw new Error('Error calling POST request: ' + err.toString());
        }

        let body;

        try {
            body = await res.json();
        }
        catch (err) {
            throw new Error('Error reading POST request response: ' + err.toString());
        }

        if (!res.ok) {
            throw new RgbRestError(res.statusText, res.status, typeof body === 'object' && body.message ? body.message : undefined);
        }

        return body;
    }

    _assembleMethodEndPointUrl(methodPath, params) {
        // Make sure that duplicate slashes that might occur in the URL (due to empty URL parameters)
        //  are reduced to a single slash so the URL used for signing is not different from the
        //  actual URL of the sent request
        return this.rootApiEndPoint + formatMethodPath(methodPath, params).replace(/\/{2,}/g,'/');
    }

}

function formatMethodPath(methodPath, params) {
    let formattedMethodPath = methodPath;

    if (typeof params === 'object' && params !== null) {
        if (typeof params.url === 'object' && params.url !== null) {
            for (let urlParam in params.url) {
                if (params.url.hasOwnProperty(urlParam)) {
                    const re = new RegExp(':' + urlParam + '\\b');
                    formattedMethodPath = formattedMethodPath.replace(re, encodeURI(params.url[urlParam]));
                }
            }
        }

        if (typeof params.query === 'object' && params.query !== null) {
            let queryStr = '';
            for (let queryParam in params.query) {
                if (params.query.hasOwnProperty(queryParam)) {
                    if (queryStr.length > 0) {
                        queryStr += '&';
                    }
                    queryStr += encodeURI(queryParam) + '=' + encodeURI(params.query[queryParam]);
                }
            }
            if (queryStr.length > 0) {
                formattedMethodPath += '?' + queryStr;
            }
        }
    }

    return formattedMethodPath;
}

function parseSuccessResponse(body) {
    if (!(typeof body === 'object' && body !== null && typeof body.status === 'string' && body.status === 'success'
            && body.hasOwnProperty('data'))) {
        throw new Error('Inconsistent successful POST request response: ' + JSON.stringify(body));
    }

    return body.data;
}

module.exports = {
    RgbRestClient,
    RgbRestError
};
